### PR TITLE
HW 04 solution by Tatsiana_Zenkovich

### PR DESCRIPTION
**Pull request by TanjaZzz**

### CODE REVIEW

File: HW04/Tatsiana_Zenkovich/spec/my_methods.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+module MyArray
+  # rubocop:disable Style/For: Prefer each over for
+  def my_each
+    if block_given?
+      for i in (0..length - 1)
+        proc.call(self[i])
+      end
+      return self
+    end
+    return to_enum(:my_each) unless block_given?
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods.rb:10]: Same thing but a little shorter
```suggestion
      return to_enum(:my_each) unless block_given?

      for i in (0..length - 1)
        proc.call(self[i])
      end
      self

```

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
+require 'my_methods.rb'
+
+describe '#my_each' do
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:3]: 💯

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  let(:a) { Array.new([1, 2, 3, 4, 5]) }
+
+  context 'a block given' do
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:6]: http://www.betterspecs.org/#contexts
`When describing a context, start its description with "when", "with" or "without".`

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        a.my_each(&block)
+      end.to yield_successive_args(1, 2, 3, 4, 5)
+    end
+
+    it 'returns the original array' do
+      expect(a.my_each {}).to eq([1, 2, 3, 4, 5])
+    end
+  end
+
+  context 'no block given' do
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:24]: 👍 for covering 2 contexts

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    it 'returns the Enumerator' do
+      expect(a.my_each).to be_an_instance_of(Enumerator)
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:26]: 👍

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
+require 'my_methods.rb'
+
+describe '#my_each' do
+  let(:a) { Array.new([1, 2, 3, 4, 5]) }
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:4]: variable name 😞 Honestly, it's a good option for `subject`
```suggestion
  subject { Array.new([1, 2, 3, 4, 5]) }
```

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  context 'a block given' do
+    it 'calls the block passed to it exact number of times' do
+      expect do |block|
+        a.my_each(&block)
+      end.to yield_control.exactly(5).times
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:10]: :open_mouth:

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    it 'yields each element to the block' do
+      expect do |block|
+        a.my_each(&block)
+      end.to yield_successive_args(1, 2, 3, 4, 5)
+    end
+
+    it 'returns the original array' do
+      expect(a.my_each {}).to eq([1, 2, 3, 4, 5])
+    end
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:21]: Can you add a test when array is empty?

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    it 'returns the Enumerator' do
+      expect(a.my_map).to be_an_instance_of(Enumerator)
+    end
+  end
+  context 'empty array is given' do
+    let(:a) { Array.new([]) }
+    it 'returns an empty array' do
+      expect(a.my_map { |x| x + 3 }).to eq([])
+    end
+  end
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:52]: Shouldn't it be under `context 'a block given'` section? 🤔

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 7 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:10]: smth wrong with it?

File: HW04/Tatsiana_Zenkovich/spec/my_methods.rb (hunk 1 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods.rb:10]: corrected, as you suggested, thanks!

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 4 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:24]: :)

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 6 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:4]: I replaced it, thank you!

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 8 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:21]: sure! done!

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 9 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:52]: right, I moved it in both cases inside a context 'when a block given', as you suggested.

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 5 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:26]: :)

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 3 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:6]: sorry, I was to pay attention to it.
corrected.

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 7 above)
Comment (vladhilko) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:10]: No, I was just surprised to see this approach 🙂

Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:10]: ah, I see)) why not)

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+  context 'when no block given' do
+    it 'returns the Enumerator' do
+      expect(subject.my_each).to be_an_instance_of(Enumerator)
+    end
+  end
+end
+# rubocop:enable Metrics/BlockLength
+describe '#my_map' do
+  subject { Array.new([1, 2, 3, 4, 5]) }
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:39]: Would be better to go by this way and just update `let` when you need 
```suggestion
  subject { Array.new(array) }
  let(:array) { [1, 2, 3, 4, 5] }
```

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    it 'returns a calculated array' do
+      expect(subject.my_map { |x| x + 3 }).to eq([4, 5, 6, 7, 8])
+    end
+
+    it 'returns an array with same length' do
+      expect(subject.my_map { |x| x + 3 }.length).to eq(subject.length)
+    end
+
+    context 'when empty array is given' do
+      subject { Array.new([]) }
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:50]: You don't need to rewrite all subject, just array inside that's why I prefer to use `let` for this case 
```suggestion
      let(:array) { [] }
```

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  end
+end
+
+describe '#my_select' do
+  subject { Array.new([1, 2, 3, 4, 5]) }
+  context 'when a block is given' do
+    it 'returns a new array which satisfy the block`s condition' do
+      expect(subject.my_select { |x| x < 3 }).to eq([1, 2])
+    end
+    it 'returns an array with at most the same length' do
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:70]: Normally, we split every `it` block with an empty line
```suggestion
    end
    
    it 'returns an array with at most the same length' do
```

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  context 'when no block given' do
+    it 'returns the Enumerator' do
+      expect(subject.my_map).to be_an_instance_of(Enumerator)
+    end
+  end
+end
+
+describe '#my_select' do
+  subject { Array.new([1, 2, 3, 4, 5]) }
+  context 'when a block is given' do
```
Comment (vladhilko) [inline HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:66]:
```suggestion
  subject { Array.new([1, 2, 3, 4, 5]) }
  
  context 'when a block is given' do
```

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 10 above)
Comment (vladhilko) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:39]: If you use the same block everywhere we can go in this way 
```suggestion
  subject { Array.new(array).my_map { |x| x + 3 } }
  # and then we can use a shorted version 
  it { is_expected.to eq ([4, 5, 6, 7, 8]) }
```

Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:39]: got it! it's no so easy readable for a newcomer, but definitely shorter) Corrected, thanks!

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 11 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:50]: corrected!

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 12 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:70]: sure, corrected it. Thanks!

File: HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb (hunk 13 above)
Comment (TanjaZzz) [reply HW04/Tatsiana_Zenkovich/spec/my_methods_spec.rb:66]: corrected it, thanks!
